pkg/operator/controller/cabundle-configmap/controller_test.go

```diff
@@ -0,0 +1,626 @@
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"github.com/stretchr/testify/assert"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+
+	"github.com/openshift/cluster-ingress-operator/pkg/manifests"
+	"github.com/openshift/cluster-ingress-operator/pkg/operator/controller"
+
+	corev1 "k8s.io/api/core/v1"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/rest"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/client/interceptor"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// getCABundleConfigMapControllerConfig initializes Config for tests.
+func getCABundleConfigMapControllerConfig() Config {
+	return Config{
+		ServiceCAConfigMapName: controller.ServiceCAConfigMapName(),
+		AdminCAConfigMapName:   controller.AdminCAConfigMapName(),
+		IngressCAConfigMapName: controller.IngressCAConfigMapName(),
+		OperatorNamespacedName: types.NamespacedName{
+			Name:      manifests.DefaultIngressControllerName,
+			Namespace: controller.DefaultOperatorNamespace,
+		},
+	}
+}
+
+// Test_New verifies that the controller's New method behaves
+// as expected.
+func Test_New(t *testing.T) {
+	mgr, err := manager.New(&rest.Config{}, manager.Options{})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	_, err = New(mgr, getCABundleConfigMapControllerConfig())
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+// Test_Reconcile verifies that the controller's Reconcile method behaves
+// correctly. The controller is expected to do the following:
+//
+//   - Create the target configmap ingress-ca-bundle in the openshift-ingress
+//     namespace with the contents read from source configmaps service-ca-bundle
+//     and admin-ca-bundle in the openshift-ingress and openshift-config namespaces
+//     respectively.
+//
+//   - Add a finalizer to the IngressController when ingress CA bundle configmap
+//     is created.  This finalizer gives the controller the opportunity to delete
+//     any configmap that it has created when the IngressController is deleted.
+//
+//   - Update the target configmap when the source configmaps are updated.
+//
+//   - Delete the target configmap and remove the finalizer if the
+//     IngressController is marked for deletion.
+//
+// The test calls the controller's Reconcile method with a fake client that has
+// suitable fake IngressController and configmap objects for the various test
+// cases to verify that the controller correctly performs the above tasks.
+func Test_Reconcile(t *testing.T) {
+	exampleTime, err := time.Parse(time.DateTime, "2006-01-02 15:04:05")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	ic := func(deleted bool, additionalFinalizers ...string) *operatorv1.IngressController {
```

Comment:
What do you think about using the existing "builder" pattern for the Ingress Controller:
https://github.com/openshift/cluster-ingress-operator/blob/7ee37ab92e459af483c3c2b1989fdb9de0864e4b/test/unit/util.go#L105

Example of usage:
https://github.com/openshift/cluster-ingress-operator/blob/7ee37ab92e459af483c3c2b1989fdb9de0864e4b/pkg/operator/controller/route-metrics/controller_test.go#L50-L54

Not required, but might be nice to get consistent. You could also add a ConfigMap builder for your configmap function.

Reply:
I will try to keep consistent.